homeassistant/helpers/intent.py

```diff
@@ -145,248 +146,543 @@ class IntentUnexpectedError(IntentError):
+    """No entities matched required states constraint."""
+
+    ASSISTANT = auto()
+    """No entities matched exposed to assistant constraint."""
+
+    INVALID_AREA = auto()
+    """Area name from constraint does not exist."""
+
+    INVALID_FLOOR = auto()
+    """Floor name from constraint does not exist."""
+
+    DUPLICATE_NAME = auto()
+    """Two or more entities matched the same name constraint and could not be disambiguated."""
+
+
+@dataclass
+class MatchTargetsConstraints:
+    """Constraints for async_match_targets."""
+
+    name: str | None = None
+    """Entity name or alias."""
+
+    area_name: str | None = None
+    """Area name, id, or alias."""
+
+    floor_name: str | None = None
+    """Floor name, id, or alias."""
+
+    domains: Collection[str] | None = None
+    """Domain names."""
+
+    device_classes: Collection[str] | None = None
+    """Device class names."""
+
+    features: int | None = None
+    """Required supported features."""
+
+    states: Collection[str] | None = None
+    """Required states for entities."""
+
+    assistant: str | None = None
+    """Name of assistant that entities should be exposed to."""
+
+    allow_duplicate_names: bool = False
+    """True if entities with duplicate names are allowed in result."""
+
+
+@dataclass
+class MatchTargetsPreferences:
+    """Preferences used to disambiguate duplicate name matches in async_match_targets."""
+
+    area_id: str | None = None
+    """Id of area to use when deduplicating names."""
+
+    floor_id: str | None = None
+    """Id of floor to use when deduplicating names."""
+
+
+@dataclass
+class MatchTargetsResult:
+    """Result from async_match_targets."""
+
+    is_match: bool
+    """True if one or more entities matched."""
+
+    no_match_reason: MatchFailedReason | None = None
+    """Reason for failed match when is_match = False."""
+
+    states: list[State] = field(default_factory=list)
+    """List of matched entity states when is_match = True."""
+
+    no_match_name: str | None = None
+    """Name of invalid area/floor or duplicate name when match fails for those reasons."""
+
+    areas: list[area_registry.AreaEntry] = field(default_factory=list)
+    """Areas that were targeted."""
+
+    floors: list[floor_registry.FloorEntry] = field(default_factory=list)
+    """Floors that were targeted."""
+
+
+class MatchFailedError(IntentError):
+    """Error when target matching fails."""
+
+    def __init__(
+        self,
+        result: MatchTargetsResult,
+        constraints: MatchTargetsConstraints,
+        preferences: MatchTargetsPreferences | None = None,
+    ) -> None:
+        """Initialize error."""
+        super().__init__()
+
+        self.result = result
+        self.constraints = constraints
+        self.preferences = preferences
+
+    def __str__(self) -> str:
+        """Return string representation."""
+        return f"result={self.result}, constraints={self.constraints}, preferences={self.preferences}"
```

Comment:
Should it contain the name of the class?
```suggestion
        return f"<MatchFailedError result={self.result}, constraints={self.constraints}, preferences={self.preferences}>"
```